instrumentation/cassandra/src/main/java/brave/cassandra/Tracing.java

```diff
@@ -0,0 +1,149 @@
+      return super.newSession(sessionId, traceType, customPayload);
+    }
+    Span span = spanFromPayload(tracer, customPayload).kind(SERVER);
+
+    // override instead of call from super as otherwise we cannot store a reference to the span
+    assert get() == null;
+    TraceState state = new ZipkinTraceState(coordinator, sessionId, traceType, span);
+    set(state);
+    sessions.put(sessionId, state);
+    return sessionId;
+  }
+
+  /** This extracts the RPC span encoded in the custom payload, or starts a new trace */
+  Span spanFromPayload(Tracer tracer, Map<String, ByteBuffer> payload) {
+    TraceContextOrSamplingFlags contextOrFlags = payload == null
+        ? TraceContextOrSamplingFlags.create(SamplingFlags.EMPTY)
+        : component.extractor().extract(payload);
+    return contextOrFlags.context() != null
+        ? tracer.joinSpan(contextOrFlags.context())
+        : tracer.newTrace(contextOrFlags.samplingFlags());
+  }
+
+  @Override protected final void stopSessionImpl() {
+    ZipkinTraceState state = (ZipkinTraceState) get();
+    if (state != null) state.incoming.finish();
+  }
+
+  @Override public final TraceState begin(String request, InetAddress client,
+      Map<String, String> parameters) {
+    ZipkinTraceState state = ((ZipkinTraceState) get());
+    Span span = state.incoming;
+    if (span.isNoop()) return state;
+
+    // request name example: "Execute CQL3 prepared query"
+    parseRequest(state, request, parameters, span);
+    // observed parameter keys include page_size, consistency_level, serial_consistency_level, query
+
+    Endpoint.Builder remoteEndpoint = Endpoint.builder().serviceName("");
+    remoteEndpoint.parseIp(client);
+    span.remoteEndpoint(remoteEndpoint.build());
+    span.start();
+    return state;
+  }
+
+  /** Defaults to the trace type. Override to use the request name as the span name */
+  protected String parseSpanName(TraceState state, String request) {
+    return state.traceType.name();
+  }
+
+  /**
+   * Override to change what data from the statement are parsed into the span representing it. By
+   * default, this sets the span name to trace type and tags {@link CassandraTraceKeys#CASSANDRA_REQUEST}
+   * and the {@link CassandraTraceKeys#CASSANDRA_SESSION_ID}.
+   *
+   * <p>If you only want to change the span name, you can override {@link #parseSpanName(TraceState,
+   * String)} instead.
+   *
+   * @see #parseSpanName(TraceState, String)
+   */
+  protected void parseRequest(TraceState state, String request, Map<String, String> parameters,
+      SpanCustomizer customizer) {
+    customizer.name(parseSpanName(state, request));
+    customizer.tag(CassandraTraceKeys.CASSANDRA_REQUEST, request);
+    customizer.tag(CassandraTraceKeys.CASSANDRA_SESSION_ID, state.sessionId.toString());
+  }
+
+  @Override protected final TraceState newTraceState(InetAddress coordinator, UUID sessionId,
+      TraceType traceType) {
+    throw new AssertionError();
```

Comment:
because this method shouldn't be called because we override what calls it. That said, if the upstream code drifts that could cease to be the case.